# Data Science Internship Challenge

Name - Janhavi Patil 

# Question 1: 

Given some sample data, write a program to answer the following: click here to access the required data set

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do    some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 

a. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 

b. What metric would you report for this dataset?

c. What is its value?

# Solution 1: Implementation

Reading and displaying the dataset csv file 

In [1]:
# importing required packages
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
pwd

'/Users/janhavi/Downloads'

Let's get some insights into data.

In [3]:
# reading the csv file
df = pd.read_csv("/Users/janhavi/Downloads/2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv")
df.head(10)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   
5         6       58      882           138            1    credit_card   
6         7       87      915           149            1           cash   
7         8       22      761           292            2           cash   
8         9       64      914           266            2          debit   
9        10       52      788           146            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11  
5  2017-03-14 15:25:01  
6  2017-03-01 21:37:57  
7   2017-03-08 2:05:38  
8  2017-03-17 20:56:50  
9  2017-03-30 21:08:26

The describe() method returns description of the data in the DataFrame

In [4]:
df.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

Getting basic information about the data

In [5]:
def all_about_my_data(df):
    print("Here is some Basic Ground Info about Data:\n")
    
    # Shape of the dataframe
    print("Number of Instances:",df.shape[0])
    print("Number of Features:",df.shape[1])
    
    # Summary Stats
    print("\nSummary Stats:")
    print(df.describe())
    
    # Checking if there are any missing values
    print("\nMissing Values:")
    print(df.isna().sum())

all_about_my_data(df)

Here is some Basic Ground Info about Data:

Number of Instances: 5000
Number of Features: 7

Summary Stats:
          order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

Missing Values:
order_id          0
shop_id           0
user_id           0
order_amount      0
total_items       0
payment_method    0
created_at        0
dtype: int64


# Observation -

1. There are 5000 rows and 7 columns for this dataset

2. In the summary stats count, mean, standard deviation, min, max and interquatile ranges are calculated

3. There are no missing values in the dataset and hence we have a clean datset

The info() function is used to print a concise summary of a DataFrame

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        5000 non-null   int64 
 1   shop_id         5000 non-null   int64 
 2   user_id         5000 non-null   int64 
 3   order_amount    5000 non-null   int64 
 4   total_items     5000 non-null   int64 
 5   payment_method  5000 non-null   object
 6   created_at      5000 non-null   object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


Getting the mean and median of order_amount

In [7]:
print('Average Order Amount:', df['order_amount'].mean(), ',',
      'Median Order Amount:', df['order_amount'].median())

Average Order Amount: 3145.128 , Median Order Amount: 284.0


Sorting the dataframe of total items in accordance with values

In [8]:
pd.DataFrame(df['total_items'].value_counts().sort_index())

total_items
1            1830
2            1832
3             941
4             293
5              77
6               9
8               1
2000           17

# Observation - 

1. Here people have got some (8 and below) pairs of shoes 
2. While there are 17 transactions where 2000 shoes have been bought. 
3. This is strange and needs to be investigated and this could possibly contibute to the high Average Order Volume. 
4. So to check if this assumption is correct or no following code is written

Filtering to get information about those orders where total items is 2000

In [9]:
df[df['total_items']==2000]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15    2017-03-07 4:00:00  
60    2017-03-04 4:00:00  
520   2017-03-02 4:00:00  
1104  2017-03-24 4:00:00  
1362  2017-03-15 4:00:00  
1436  2017-03-11 4:00:00  
1562  2017-03-19 4:00:00  
1602  2017-03-17 4:00:00  
2153  2017-03-12 4:00:00  
2297  2017-03-07 4:00:00  
2835  2017-03-28 4:00:00  
2969  2017-03-28 4:00:00  
3332  2017-03-24 4:00:00  
4056  2017-03-28 4:00:00  
4646  2017-03-02 4:00:00  
4868  2017-03-22 4:00:00  
4882  2017-03-25 4:00:00

# Observation - 

1. The transaction with total_items equal to 2000 are happening at the same shop by the same customer at the same time. 
2. There can be some fraudulent activity with the customer or something is wrong with the shop which is shop 42

To check that the following code is written

Filtering to get information about those orders where shop ID is 42

In [10]:
df[df['shop_id']==42]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
40          41       42      793           352            1    credit_card   
60          61       42      607        704000         2000    credit_card   
308        309       42      770           352            1    credit_card   
409        410       42      904           704            2    credit_card   
520        521       42      607        704000         2000    credit_card   
834        835       42      792           352            1           cash   
835        836       42      819           704            2           cash   
938        939       42      808          1056            3    credit_card   
979        980       42      744           352            1          debit   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1436      1437       42      607        704000         2000    credit_card   
1471      1472       42      907          1408            4          debit   
1512      1513       42      946           352            1          debit   
1520      1521       42      756           704            2          debit   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
1911      1912       42      739           704            2           cash   
1929      1930       42      770           352            1    credit_card   
2003      2004       42      934           704            2           cash   
2018      2019       42      739           352            1          debit   
2053      2054       42      951           352            1          debit   
2153      2154       42      607        704000         2000    credit_card   
2273      2274       42      747           704            2          debit   
2297      2298       42      607        704000         2000    credit_card   
2491      2492       42      868           704            2          debit   
2609      2610       42      868           704            2          debit   
2766      2767       42      970           704            2    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
2987      2988       42      819          1056            3           cash   
3332      3333       42      607        704000         2000    credit_card   
3513      3514       42      726          1056            3          debit   
3651      3652       42      830           352            1    credit_card   
3697      3698       42      839           352            1          debit   
3903      3904       42      975           352            1          debit   
3998      3999       42      886           352            1          debit   
4056      4057       42      607        704000         2000    credit_card   
4231      4232       42      962           352            1           cash   
4294      4295       42      859           704            2           cash   
4326      4327       42      788           704            2          debit   
4421      4422       42      736           704            2    credit_card   
4625      4626       42      809           352            1    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4745      4746       42      872           352            1          debit   
4767      4768       42      720           704            2    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42     

# Observation -

All transactions associated with User_id 607 are outliers

So tghe following code is written to drop these values from our dataset and to re calculate mean and median to know if there is still any outlier in the data

In [11]:
subset_df = df[df['user_id']!=607]
subset_df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [12]:
print('Average Order Amount:', subset_df['order_amount'].mean(), ',',
      'Median Order Amount:', subset_df['order_amount'].median())

Average Order Amount: 754.0919125025085 , Median Order Amount: 284.0


# Observation -

1. There is still some discripancy 
2. So calculated the price per pair of shoes in our dataset

In [13]:
subset_df['price_per_item'] = subset_df['order_amount']/subset_df['total_items']

<ipython-input-13-6c2d5353d879>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset_df['price_per_item'] = subset_df['order_amount']/subset_df['total_items']


In [14]:
pd.DataFrame(subset_df['price_per_item'].describe())

price_per_item
count      4983.00000
mean        387.86474
std        2446.12562
min          90.00000
25%         133.00000
50%         153.00000
75%         168.00000
max       25725.00000

# Observation - 

1. The max value is coming to be 25725
2. Which is suspectful for a pair of sneakers 

In [15]:
pd.DataFrame(subset_df['price_per_item'].value_counts().sort_index())

price_per_item
90.0                 42
94.0                 55
101.0                42
111.0                40
112.0               124
114.0                59
116.0                58
117.0                77
118.0               118
122.0                50
127.0                52
128.0                93
129.0               142
130.0               150
131.0                97
132.0                46
133.0               151
134.0                86
136.0               105
138.0                59
140.0                55
142.0               197
144.0                39
145.0                92
146.0                89
147.0                53
148.0               101
149.0                52
153.0               256
154.0                54
155.0                42
156.0               197
158.0                94
160.0               206
161.0               101
162.0                48
163.0               122
164.0               161
165.0                58
166.0                43
168.0                39
169.0                54
171.0                48
172.0                35
173.0                99
176.0               152
177.0               148
178.0               109
181.0               112
184.0                49
187.0               105
190.0                35
193.0                44
195.0                54
196.0                61
201.0                53
352.0                34
25725.0              46

# Observation - 

1. Roughly all prices are below 400
2. Except for the last value which says 46 orders were sold for 25725$ each
3. This is hence an outlier

In [16]:
normal_orders = subset_df[subset_df['price_per_item'] < 400 ]

outlier_orders = subset_df[subset_df['price_per_item'] == 25725]

In [17]:
outlier_orders

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  price_per_item  
160    2017-03-12 5:56:57         25725.0  
490   2017-03-26 17:08:19         25725.0  
493   2017-03-16 21:39:35         25725.0  
511    2017-03-09 7:23:14         25725.0  
617   2017-03-18 11:18:42         25725.0  
691   2017-03-27 22:51:43         25725.0  
1056  2017-03-15 10:16:45      

# Observation - 

1. Shop 78 has 0 orders under 400 per pair of shoes
2. Shop 78 has 46 orders priced at 25,725 per pair of shoes

In [18]:
len(normal_orders[normal_orders['shop_id']==78]), len(outlier_orders[outlier_orders['shop_id']==78])

(0, 46)

Creating a final dataframe by removing these outlier values

In [19]:
df = df[df['user_id']!=607]
df = df[df['shop_id']!=78]
df

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[4937 rows x 7 columns]

In [20]:
df['order_amount'].mean(), df['order_amount'].median()

(302.58051448247926, 284.0)

# Detailed Answers:

# Q1. Think about what could be going wrong with our calculation. 

Answer:

On performing basic exploratory data analysis came to know that there are outliers persent which are observations that lie on an abnormal distance from other values in a random sample from a population. After writing filtering code on the dataset user ID 607 anf Shop ID 78 were found to have some suspicious activity going on and were the outliers in the dataset. This was contributing to the errors in the overall calculations. 

# Q2. What metric would you report for this dataset?

Answer:

The metric to be used for this dataset would be a 'Median Order Value' since there are outliers present in our data-set. But on dropping the outliers 'Mean Order Value' can be used


# Q3. What is its value?

Answer:

284.0 (Median Order Value)

After removing the outliers,

302.58 - Mean Order Value 